iseq.c

```diff
@@ -248,230 +245,155 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
-{
-    if (ISEQ_BODY(iseq)) {
-        struct rb_iseq_constant_body *body = ISEQ_BODY(iseq);
-
-        body->variable.coverage = rb_gc_location(body->variable.coverage);
-        body->variable.pc2branchindex = rb_gc_location(body->variable.pc2branchindex);
-        body->variable.script_lines = rb_gc_location(body->variable.script_lines);
-        body->location.label = rb_gc_location(body->location.label);
-        body->location.base_label = rb_gc_location(body->location.base_label);
-        body->location.pathobj = rb_gc_location(body->location.pathobj);
-        if (body->local_iseq) {
-            body->local_iseq = (struct rb_iseq_struct *)rb_gc_location((VALUE)body->local_iseq);
-        }
-        if (body->parent_iseq) {
-            body->parent_iseq = (struct rb_iseq_struct *)rb_gc_location((VALUE)body->parent_iseq);
-        }
-        if (body->mandatory_only_iseq) {
-            body->mandatory_only_iseq = (struct rb_iseq_struct *)rb_gc_location((VALUE)body->mandatory_only_iseq);
-        }
-        if (body->call_data) {
-            for (unsigned int i=0; i<body->ci_size; i++) {
-                struct rb_call_data *cds = body->call_data;
-                if (!SPECIAL_CONST_P((VALUE)cds[i].ci)) {
-                    cds[i].ci = (struct rb_callinfo *)rb_gc_location((VALUE)cds[i].ci);
-                }
-                cds[i].cc = (struct rb_callcache *)rb_gc_location((VALUE)cds[i].cc);
-            }
-        }
-        VALUE *original_iseq = ISEQ_ORIGINAL_ISEQ(iseq);
-        rb_iseq_each_value(iseq, update_each_insn_value, (void *)original_iseq);
-
-        if (body->param.flags.has_kw && ISEQ_COMPILE_DATA(iseq) == NULL) {
-            int i, j;
-
-            i = body->param.keyword->required_num;
-
-            for (j = 0; i < body->param.keyword->num; i++, j++) {
-                VALUE obj = body->param.keyword->default_values[j];
-                if (!UNDEF_P(obj)) {
-                    body->param.keyword->default_values[j] = rb_gc_location(obj);
-                }
-            }
-        }
-
-        if (body->catch_table) {
-            struct iseq_catch_table *table = body->catch_table;
-            unsigned int i;
-            for (i = 0; i < table->size; i++) {
-                struct iseq_catch_table_entry *entry;
-                entry = UNALIGNED_MEMBER_PTR(table, entries[i]);
-                if (entry->iseq) {
-                    entry->iseq = (rb_iseq_t *)rb_gc_location((VALUE)entry->iseq);
+                    iseq_scan_bits(i, bits, code, original_iseq);
                 }
             }
         }
-#if USE_MJIT
-        mjit_update_references(iseq);
-#endif
-#if USE_YJIT
-        rb_yjit_iseq_update_references(body->yjit_payload);
-#endif
     }
 }
 
-static VALUE
-each_insn_value(void *ctx, VALUE obj)
-{
-    rb_gc_mark_movable(obj);
-    return obj;
-}
-
 void
-rb_iseq_mark(const rb_iseq_t *iseq)
+rb_iseq_mark_and_update(rb_iseq_t *iseq, bool reference_updating)
```

Comment:
Adding the `reference_updating` parameter was due to a technical limitation of iseqs since it needs to read the `rb_callinfo` in the `ci` (since it might not be a Ruby object, so we can't mark it in this case) and `rb_callcache` in the `cc` (since it will be updated if it has been invalidated, but we can't read the object if it has been moved).

I think one solution is to make both of these data types (`rb_callinfo` and `rb_callcache`) store their data via malloc and use their Ruby objects as plain wrappers over the malloc'd region. The downside is of course the performance (and memory) overhead of malloc while allocating these objects. But this way we can always read the data without worrying about compaction. Another option would be to make them non-Ruby objects and always allocated through malloc. That way we don't have to mark them but we'll have to figure out how to ensure that it doesn't leak memory. So I'm not sure what the best solution here is.

I've tried to keep the code wrapped in `reference_updating` small, and only used it in the two cases above when it's absolutely necessary.